### PR TITLE
Allow empty features in time series pipelines

### LINKED ISSUE #1517: Allow X to be None in Time Series Pipelines
In previous EvalML pipelines we have always made the assumption that `X` cannot be `None`. However, with the addition of the `TimeSeriesBaselineRegressor` and plans to add more estimators that rely only on `y` (such as ARIMA, PROPHET) we can revise this assumption for TimeSeriesRegressionPipelines. 

This issue keeps track of:

1. Removing the `ValueError` from `TimeSeriesRegressionPipeline.fit` and accommodating for an empty `X`
2. Removing the overridden `fit` method in `TimeSeriesBaseLineRegressionPipeline` as that was put in as a patch to solve this problem

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fixes #1517.

@dsherry and @bchen1116 Can't tag you as reviewers but please take a look when you can!

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: evalml/pipelines/components/estimators/regressors/time_series_baseline_regressor.py [hunk 1]
```diff
@@ -52,12 +53,8 @@ def fit(self, X, y=None):
         return self
 
     def predict(self, X, y=None):
-        if X is None:
```
Comment (freddyaboulton) [inline evalml/pipelines/components/estimators/regressors/time_series_baseline_regressor.py:55]: Deleting these because it's not needed

File: evalml/pipelines/regression/time_series_baseline_regression.py [hunk 2]
```diff
@@ -1,64 +1,7 @@
...
-    pad_with_nans
-)
 
 
 class TimeSeriesBaselineRegressionPipeline(TimeSeriesRegressionPipeline):
     """Baseline Pipeline for time series regression problems."""
     _name = "Time Series Baseline Regression Pipeline"
     component_graph = ["Time Series Baseline Regressor"]
-
-    def fit(self, X, y):
```
Comment (freddyaboulton) [inline evalml/pipelines/regression/time_series_baseline_regression.py:18]: This code was needed to accept the `X=None` case so we can delete it now! This means adding baselines for time series classification will be easy now since we can just reuse the time series baseline regressor and not worry about implementing pipeline `fit` or `predict` !

File: evalml/pipelines/components/estimators/estimator.py [hunk 3]
```diff
@@ -21,6 +21,8 @@ class Estimator(ComponentBase):
 
     To see some examples, check out the definitions of any Estimator component.
     """
+    # We can't use the inspect module to dynamically determine this because of issue 1582
+    predict_uses_y = False
```
Comment (freddyaboulton) [inline evalml/pipelines/components/estimators/estimator.py:25]: We need a way of knowing whether or not to pass `y` as an argument to `predict` or `predict_proba` since one (more in the future) estimator allows this and others don't. I think this is the lowest-touch way of doing this.

One alternative is to list `y` as an argument for all estimators. However, I think that would be confusing for users to list `y` if it's not used because they might think our estimators are "cheating" by looking at the target value. Maybe I'm just being paranoid though.

**Review by chukarsten — APPROVED** [review]
This seems conceptually sound.  Definitely a fan of removing unused code and simplification.

**Review by bchen1116 — APPROVED** [review]
LGTM!

File: evalml/pipelines/time_series_classification_pipelines.py [hunk 4]
```diff
@@ -61,21 +60,41 @@ def fit(self, X, y):
...
+
+    def _estimator_predict(self, features, y):
+        """Get estimator predictions.
+
+        This helper passes y as an argument if needed by the estimator.
+        """
+        return self._estimator_predict_helper(features, y, return_probabilities=False)
+
+    def _estimator_predict_proba(self, features, y):
+        """Get estimator prediced probabilities.
```
Comment (angela97lin) [inline evalml/pipelines/time_series_classification_pipelines.py:87]: Typo here :d

File: evalml/pipelines/time_series_classification_pipelines.py [hunk 5]
```diff
@@ -113,14 +132,15 @@ def predict_proba(self, X, y=None):
...
+        proba = self._estimator_predict_proba(features_no_nan, y_encoded)
         proba.columns = self._encoder.classes_
         return pad_with_nans(proba, max(0, features.shape[0] - proba.shape[0]))
 
     def _compute_predictions(self, X, y, objectives):
         """Compute predictions/probabilities based on objectives."""
-        y_predicted = None
-        y_predicted_proba = None
+        y_predicted = pd.Series([])
+        y_predicted_proba = pd.DataFrame([])
```
Comment (angela97lin) [inline evalml/pipelines/time_series_classification_pipelines.py:143]: I may be confused but why this change? This'll always return a valid prediction value, even if its empty?

File: evalml/pipelines/time_series_classification_pipelines.py (hunk 4 above)
Comment (freddyaboulton) [reply evalml/pipelines/time_series_classification_pipelines.py:87]: Thank you!

File: evalml/pipelines/time_series_classification_pipelines.py (hunk 5 above)
Comment (freddyaboulton) [reply evalml/pipelines/time_series_classification_pipelines.py:143]: `drop_rows_with_nans` (which is called in `score`) needs a pandas data type and not `None`. I changed `drop_rows_with_nans` to accept None and so I changed these back to `None`.

File: evalml/tests/pipeline_tests/test_time_series_pipeline.py [hunk 6]
```diff
@@ -319,6 +322,37 @@ class Pipeline(pipeline_class):
...
+        label_map = {0: "good", 1: "bad", 2: "best"}
+        y = pd.Series(y).map(lambda label: label_map[label])
+        expected_unique_values = {"good", "bad", "best"}
+        objectives = ["MCC Multiclass", "Log Loss Multiclass"]
+
+    if use_none_X:
+        X = None
+
+    pl.fit(X, y)
+    if expected_unique_values:
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/tests/pipeline_tests/test_time_series_pipeline.py:350]: Yep, removing the if statement!

File: evalml/pipelines/time_series_classification_pipelines.py [hunk 7]
```diff
@@ -61,21 +60,41 @@ def fit(self, X, y):
...
+        This helper passes y as an argument if needed by the estimator.
+        """
+        return self._estimator_predict_helper(features, y, return_probabilities=False)
+
+    def _estimator_predict_proba(self, features, y):
+        """Get estimator predicted probabilities.
+
+        This helper passes y as an argument if needed by the estimator.
+        """
+        return self._estimator_predict_helper(features, y, return_probabilities=True)
```
Comment (dsherry) [inline evalml/pipelines/time_series_classification_pipelines.py:91]: @freddyaboulton could we do this
```python
def _estimator_predict(self, features, y):
        """Get estimator predictions.
        This helper passes y as an argument if needed by the estimator.
        """
        y_arg = None
        if self.estimator.predict_uses_y:
            y_arg = y
        return self.estimator.predict(features, y_arg, return_probabilities=False)
```
And similar for predict_proba?

File: evalml/pipelines/time_series_classification_pipelines.py [hunk 8]
```diff
@@ -172,21 +187,22 @@ def threshold(self, value):
...
             if not objective.is_defined_for_problem_type(self.problem_type):
                 raise ValueError(f"Objective {objective.name} is not defined for time series binary classification.")
 
         if self.threshold is None:
-            predictions = self.estimator.predict(features_no_nan)
+            predictions = self._estimator_predict(features_no_nan, y_encoded)
         else:
-            ypred_proba = self.estimator.predict_proba(features_no_nan).iloc[:, 1]
+            proba = self._estimator_predict_proba(features_no_nan, y_encoded)
+            proba = proba.iloc[:, 1]
```
Comment (dsherry) [inline evalml/pipelines/time_series_classification_pipelines.py:201]: Why is this necessary? Not covered by `_estimator_predict_proba`?

File: evalml/tests/pipeline_tests/test_time_series_pipeline.py [hunk 9]
```diff
@@ -319,6 +322,36 @@ class Pipeline(pipeline_class):
...
+        objectives = ['MCC Binary', "Log Loss Binary"]
+    elif pl.problem_type == ProblemTypes.TIME_SERIES_MULTICLASS:
+        X, y = X_y_multi
+        label_map = {0: "good", 1: "bad", 2: "best"}
+        y = pd.Series(y).map(lambda label: label_map[label])
+        expected_unique_values = {"good", "bad", "best"}
+        objectives = ["MCC Multiclass", "Log Loss Multiclass"]
+
+    if use_none_X:
+        X = None
```
Comment (dsherry) [inline evalml/tests/pipeline_tests/test_time_series_pipeline.py:347]: Do we already have coverage elsewhere for the case when `use_none_X` is `True`? I.e. do we need `@pytest.mark.parametrize("use_none_X", [True, False])`?

File: evalml/utils/gen_utils.py [hunk 10]
```diff
@@ -401,19 +401,24 @@ def _not_nan(pd_data):
     return mask
 
 
-def drop_rows_with_nans(pd_data_1, pd_data_2):
+def drop_rows_with_nans(*pd_data):
     """Drop rows that have any NaNs in both pd_data_1 and pd_data_2.
```
Comment (dsherry) [inline evalml/utils/gen_utils.py:405]: Nit-pick: update this docstring too

File: evalml/pipelines/regression/time_series_baseline_regression.py (hunk 2 above)
Comment (dsherry) [reply evalml/pipelines/regression/time_series_baseline_regression.py:18]: Ooh that's great!

File: evalml/pipelines/time_series_classification_pipelines.py [hunk 11]
```diff
@@ -61,21 +60,41 @@ def fit(self, X, y):
         y = self._encode_targets(y)
 
         X_t = self._compute_features_during_fit(X, y)
-        if X_t.empty:
-            raise RuntimeError("Pipeline computed empty features during call to .fit. This means "
-                               "that either 1) you passed in X=None to fit and don't have a DelayFeatureTransformer "
-                               "in your pipeline or 2) you do have a DelayFeatureTransformer but gap=0 and max_delay=0. "
-                               "Please add a DelayFeatureTransformer or change the values of gap and max_delay")
```
Comment (dsherry) [inline evalml/pipelines/time_series_classification_pipelines.py:68]: Ah got it, great that we can eliminate this!

File: evalml/pipelines/time_series_classification_pipelines.py [hunk 12]
```diff
@@ -146,13 +166,8 @@ def score(self, X, y, objectives):
         y_encoded = self._encode_targets(y)
         y_shifted = y_encoded.shift(-self.gap)
         y_pred, y_pred_proba = self._compute_predictions(X, y, objectives)
-        non_nan_mask = _get_rows_without_nans(y_shifted, y_pred, y_pred_proba)
-        if y_pred is not None:
-            y_pred = y_pred.iloc[non_nan_mask]
-        if y_pred_proba is not None:
-            y_pred_proba = y_pred_proba.iloc[non_nan_mask]
-        y_labels = y_shifted.iloc[non_nan_mask]
```
Comment (dsherry) [inline evalml/pipelines/time_series_classification_pipelines.py:154]: Is it necessary to remove the `iloc` stuff from where this code got moved to?

**Review by dsherry — APPROVED** [review]
I left a suggestion in the time series classification pipeline, and a couple questions, otherwise 🚢

File: evalml/pipelines/time_series_classification_pipelines.py (hunk 8 above)
Comment (freddyaboulton) [reply evalml/pipelines/time_series_classification_pipelines.py:201]: Yea, `_estimator_predict_proba` returns the probabilities for all classes. For binary pipelines in `predict`, we threshold the predictions for the "positive" class if the threshold is not None. Hence the need for `.iloc[:, 1]`.

File: evalml/pipelines/time_series_classification_pipelines.py (hunk 12 above)
Comment (freddyaboulton) [reply evalml/pipelines/time_series_classification_pipelines.py:154]: The calls to `iloc` are now happening in `drop_rows_with_nans`!

File: evalml/tests/pipeline_tests/test_time_series_pipeline.py (hunk 9 above)
Comment (freddyaboulton) [reply evalml/tests/pipeline_tests/test_time_series_pipeline.py:347]: Yes I think we do for now. In the other tests with `use_none_X`, we always have a `DelayedFeaturesTransformer` in the pipeline. I think it's important to have coverage when that's not the case. 

That being said, we have coverage for this case for ts regression pipelines in `test_time_series_baseline_regression.py`. In #1580 , I plan on extending the time baseline regressor to classification problems so in that case, I think we can move fold in this coverage into the tests for the baseline.

File: evalml/utils/gen_utils.py (hunk 10 above)
Comment (freddyaboulton) [reply evalml/utils/gen_utils.py:405]: Thank you!

File: evalml/pipelines/time_series_classification_pipelines.py (hunk 7 above)
Comment (freddyaboulton) [reply evalml/pipelines/time_series_classification_pipelines.py]: Yes we can! I originally thought we couldn't because some estimators like `XGBoost` don't have `y` as an argument in their predict signature but the metaclass adds it!
